### PR TITLE
Use std::atomic for IsChainNearlySyncd()

### PR DESCRIPTION
**Pull request by ptschip**
This function gets called frequently  at the time we are looking
for a peer to download an xthin from.  This should give a small
performance boost.

### CODE REVIEW

File: src/unlimited.cpp [hunk 1]
```diff
@@ -1053,21 +1054,23 @@ UniValue settrafficshaping(const UniValue& params, bool fHelp)
...
-      if(chainActive.Height() < pindexBestHeader->nHeight - 2)
-        fIsChainNearlySyncd = false;
-      else
-        fIsChainNearlySyncd = true;
-      }
+    bool flag = fIsChainNearlySyncd.load();
+    LOCK(cs_main);
+    if (!pindexBestHeader) {
+        // Not nearly synced if we don't have any blocks!
+        while (flag && !fIsChainNearlySyncd.compare_exchange_weak(flag, false));
```
Comment (gandrewstone) [inline src/unlimited.cpp]: I think this code would work, but AFAIK the purpose of compare_exchange_xxx is handle a situation where the code needs to use the current value of the atomic.  For example for a read, add, write, you'd do:
int current = val.load();
while (!fIsChainNearlySyncd.compare_exchange_weak(val, val+5));

But if you just need to set the value of the object, irrespective of its current value, then use the "store" API: http://en.cppreference.com/w/cpp/atomic/atomic/store

So in a nutshell, I think it would be a lot more clear to replace lines 1061, 1065 with
fIsChainNearlySyncd.store(false)
and 1067 with .store(true)

**Review by gandrewstone — CHANGES_REQUESTED** [review]

File: src/unlimited.cpp (hunk 1 above)
Comment (deadalnix) [reply src/unlimited.cpp]: I'm not sure I understand the whole flow here. Basically, we want fIsChainNearlySyncd to be true if are within 2 block of the best header we know of.

I'm not sure when this function is called so it's difficult to assert correctness or not.

**Comment by ptschip** [comment]
@deadalnix On 15/02/2017 8:50 AM, deadalnix wrote:
>
> *@deadalnix* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/unlimited.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/273>:
>
> > @@ -1053,21 +1054,23 @@ UniValue settrafficshaping(const UniValue& params, bool fHelp)
>  // This way we avoid having to lock cs_main so often which tends to be a bottleneck.
>  void IsChainNearlySyncdInit() 
>  {
> -    LOCK2(cs_main, cs_ischainnearlysyncd);
> -    if (!pindexBestHeader) fIsChainNearlySyncd = false;  // Not nearly synced if we don't have any blocks!
> -    else
> -      {
> -      if(chainActive.Height() < pindexBestHeader->nHeight - 2)
> -        fIsChainNearlySyncd = false;
> -      else
> -        fIsChainNearlySyncd = true;
> -      }
> +    bool flag = fIsChainNearlySyncd.load();
> +    LOCK(cs_main);
> +    if (!pindexBestHeader) {
> +        // Not nearly synced if we don't have any blocks!
> +        while (flag && !fIsChainNearlySyncd.compare_exchange_weak(flag, false));
>
> I'm not sure I understand the whole flow here. Basically, we want
> fIsChainNearlySyncd to be true if are within 2 block of the best
> header we know of.
>
yes, the logic hasn't changed it's just that i wanted to use atomic
here.  So anyway, yes what do here is that this function gets called
every time a new header arrives.  So we can then
update the flat do say whether we are nearly sync'd to the chain or
not.  Based on this, the other function IsChainNearlySyncd() is called
(when we want to determine wheter to ask for an xthin or a regular block).
>
> I'm not sure when this function is called so it's difficult to assert
> correctness or not.it
>
look for it in main.cpp.  also you 'll see reference to
IsChainNearlySyncd() in the requestManager.cpp

**Comment by deadalnix** [comment]
So first, @gandrewstone is right, you don't need CAS operations here, they don't bring you anything.

The more general problem I see here is that you don't have any kind of synchronization between what read that variable and what writes it. It looks like to me that it should probably not be a global at all but be computed when a block is found and passed down the the code that validate.

**Comment by ptschip** [comment]
@gandrewstone  i made those requested changes...

**Review by gandrewstone — APPROVED** [review]
